gcc/rust/checks/errors/borrowck/rust-function-collector.h

```diff
@@ -0,0 +1,198 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef RUST_HIR_FUNCTION_COLLECTOR_H
+#define RUST_HIR_FUNCTION_COLLECTOR_H
+
+#include "rust-hir-item.h"
+#include "rust-hir-visitor.h"
+#include "rust-hir.h"
+
+#include <vector>
+
+namespace Rust {
+
+// Collects all HIR items eligible for borrow checking.
+class FunctionCollector : public HIR::HIRFullVisitor
+{
+  std::vector<HIR::Function *> functions;
+  std::vector<HIR::ClosureExpr *> closures;
+
+public:
+  void go (HIR::Crate &crate) { visit_all (crate.get_items ()); }
+
+  [[nodiscard]] const std::vector<HIR::Function *> &get_functions () const
+  {
+    return functions;
+  }
+
+  [[nodiscard]] const std::vector<HIR::ClosureExpr *> &get_closures () const
+  {
+    return closures;
+  }
+
+protected:
+  template <typename T> void visit_all (std::vector<std::unique_ptr<T>> &items)
+  {
+    for (std::unique_ptr<T> &item : items)
+      {
+	item->accept_vis (*this);
+      }
+  }
+
+  void visit (HIR::Function &function) override
+  {
+    functions.push_back (&function);
+    function.get_definition ()->accept_vis (*this);
+  }
+
+  void visit (HIR::ClosureExpr &closure) override
+  {
+    closures.push_back (&closure);
+    closure.get_expr ()->accept_vis (*this);
+  }
```

Comment:
the implementation for these functions should live in a `.cc` file. This can be done in a separate PR

Reply:
The header file is only used in one place, so it seamed like useless separations. I only created a cc file if there was a reason. Most of my implementation is in header files. I like it that way more, but it is up to you.

Reply:
putting it in headers slows down the compilation a lot sadly, since we constantly do a lot of includes. it does reflect on our compile time if the implementation lives in headers vs source files, so we've been trying to keep it that way. especially for "complex" functions, which could get modified later on like these. a simple getter or constructor can live in a header, that's no problem